カレンダーのプログラム(JavaScript)を作成

sjabcdefin opened the pull request:
カレンダーのプログラム(JavaScript)を作成いたしました。
ご確認お願いいたします。

sjabcdefin committed:
Create: Create calendar program

カレンダープログラムを作成。calコマンドと同じ見た目にし、-mで月を、-yで年を指定可能とする。

sjabcdefin committed:
Update: Change file name

ファイル名をcalendar.jsからcal.jsに変更

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+
+const params = minimist(process.argv.slice(2));
+const month =
+  params["m"] === undefined ? new Date().getMonth() : params["m"] - 1;
```
`params["m"]` の部分についてですが、オブジェクトにアクセスするときは基本的にドット記法を用いてください。

sjabcdefin replied:
ドット記法```params.m```, ```params.y```に修正いたしました。

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/22ec1969e4bb112fd815a58edf7e4a679e4b6624

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+const params = minimist(process.argv.slice(2));
+const month =
+  params["m"] === undefined ? new Date().getMonth() : params["m"] - 1;
+const year = params["y"] === undefined ? new Date().getFullYear() : params["y"];
```
特定の値が `undefined` のときにこのような処理を行いたいのであれば、別の演算子を使えばもっと簡潔に書けるはずです。

sjabcdefin replied:
??演算子を使用するように修正いたしました。
```params.m```や```params.y```が```undefined```の場合、??の右側の値を返し、
```params.m```や```params.y```が```undefined```出ない場合、??の左側の値を返す。

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/22ec1969e4bb112fd815a58edf7e4a679e4b6624

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+
+const params = minimist(process.argv.slice(2));
+const month =
+  params["m"] === undefined ? new Date().getMonth() : params["m"] - 1;
+const year = params["y"] === undefined ? new Date().getFullYear() : params["y"];
```
`new Date()` を 2 行に渡って評価すると `month` と `year` に意図しない値が設定される可能性があります。例えば`month` の行は `2024-12-31 23:59:59` に評価されるが、`year` の行は `2025-01-01 00:00:00` に評価されるようなケースです。

sjabcdefin replied:
変数```today```に```new Date()```を設定し、```month```, ```year```には、```today```から取得した月、年をそれぞれ設定するように修正しました。

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/22ec1969e4bb112fd815a58edf7e4a679e4b6624

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+  params["m"] === undefined ? new Date().getMonth() : params["m"] - 1;
+const year = params["y"] === undefined ? new Date().getFullYear() : params["y"];
+
+const start_of_month = new Date(year, month);
```
識別子には came case を用いてください。

sjabcdefin replied:
```start_of_month```→```startDate```, ```end_of_month```→```endDate```に修正いたしました。
(https://github.com/sjabcdefin/js-practices/pull/2#discussion_r1889359522 の指摘も反映)

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/421a65c9f1488ac22711104864c4a408c88e16d8

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+const start_of_month = new Date(year, month);
+const end_of_month = new Date(year, start_of_month.getMonth() + 1, 0);
+
+console.log(`      ${month + 1}月 ${year}     `);
```
右側の空白はいらないです。つけるにしても月の表示幅は可変なので、この実装だと表示幅が固定にならないです。

sjabcdefin replied:
```${year}```の右側の空白は削除いたしました。

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/10245a190559df15d14ed6e078bf7d0751c05643

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+
+console.log(`      ${month + 1}月 ${year}     `);
+console.log("日 月 火 水 木 金 土");
+process.stdout.write(" ".repeat(3 * start_of_month.getDay()));
```
空白の総数としては `3 * start_of_month.getDay()` かもしれませんが、意味的には `<空白 3 つ> *  start_of_month.getDay()` だと思います。それを表現するようなコードにしてください。

sjabcdefin replied:
```"   ".repeat(startDate.getDay())```に修正いたしました。

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/fda2227d860e39444a74c1c2e831b51bf7f1af28

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+process.stdout.write(" ".repeat(3 * start_of_month.getDay()));
+
+for (
+  const day = new Date(start_of_month);
```
`start_of_month`, `end_of_month` と `day` はいずれも Date オブジェクトが格納される変数なのに命名に一貫性がないです。

sjabcdefin replied:
```start_of_month```→```startDate```
```end_of_month``` → ```endDate```
```day``` → ```date```
に修正いたしました。

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/421a65c9f1488ac22711104864c4a408c88e16d8

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+  day <= end_of_month;
+  day.setDate(day.getDate() + 1)
+) {
+  process.stdout.write(`${String(day.getDate()).padStart(2, " ")} `);
```
この実装だと土曜日の日付の後ろに余分な空白が出力されます。

sjabcdefin replied:
土曜日の日付の後ろに余分な表示がされないよう、
土曜日の場合は改行を出力、土曜日以外の場合に、空白を出力するように修正いたしました。
```
  if (date.getDay() === 6) {
    process.stdout.write("\n");
  } else {
    process.stdout.write(" ");
  }
```

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/a7dcd307ade6f9f3ab5601ad9c32e60d5e6247a5

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,27 @@
+  day.setDate(day.getDate() + 1)
+) {
+  process.stdout.write(`${String(day.getDate()).padStart(2, " ")} `);
+  if (day.getDay() === 6) {
+    process.stdout.write("\n");
+  }
+}
+process.stdout.write("\n\n");
```
この実装だと月末の日付が土曜日のときに空行が余分に出力されます。

sjabcdefin replied:
ご指摘いただいた「月末の日付が土曜日の場合」のほかに、
2000年4月のような日付表示部分が6行の場合も、空行が余分に出力されていました。
```
❯ ./cal.js -m 4 -y 2000
      4月 2000
日 月 火 水 木 金 土
                   1
 2  3  4  5  6  7  8
 9 10 11 12 13 14 15
16 17 18 19 20 21 22
23 24 25 26 27 28 29
30


❯ cal 4 4000
      4月 4000
日 月 火 水 木 金 土
                   1
 2  3  4  5  6  7  8
 9 10 11 12 13 14 15
16 17 18 19 20 21 22
23 24 25 26 27 28 29
30
```
(私の環境では、)日付表示部分は、空行含め7行であったため、以下のように修正し、余分な空行は出力されないように修正いたしました。
```
for (let lines = numberOfLines; lines < 6; lines++) {
  process.stdout.write("\n");
}
```

```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/2ab578a66c15238765f5ca7dc9e3194a3641b37b

sjabcdefin replied:
```eslint-plugin-prettier```をインストール、```eslint.config.js```の修正をし、ESLint実行時にPrettierのフォーマット違反を検出できるようにいたしました。

sjabcdefin committed:
Update: Modify the method to obtain the year and month to display.

表示する年、月の取得方法を修正。
- param["m"],param["y"]について、ドット記法に修正
- undefinedの場合の判定に、??演算子を使用するように修正
- 変数month, yearについて、new Date()を変数todayに設定し、todayからmonth, yearをそれぞれ取得するように修正

sjabcdefin committed:
Update: Modify the variable name of Date object.

Dateオブジェクトが格納される変数名を変更。
- 変数(startDate, endDate)について、キャメルケースを使用するように修正
- 変数名start_of_monthをstartDate, end_of_monthをendDate, dayをdateに修正

sjabcdefin committed:
Update: Delete the blank on the right side of the year and month.

年月右側の空白を削除

sjabcdefin committed:
Update: Modify the method to calculate the total number of blank before the day is displayed.

カレンダーの最初の日付表示前の空白総数の算出方を変更。

sjabcdefin committed:
Update: Modify so that the blank behind Saturday isn't display.

土曜日の日付の後ろに余分は空白が表示されないよう修正。

sjabcdefin committed:
Update: Modify so that extra blank lines are not output.

最終行に余分な空行が出力されないように修正。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+const year = params.y ?? today.getFullYear();
+
+const startDate = new Date(year, month - 1);
+const endDate = new Date(year, startDate.getMonth() + 1, 0);
```
年に関しては `year` が使われているのに月に関しては `startDate.getMonth()` が使われていて一貫性がないです。`startDate` からでないと参照できないような値でないのであれば、`year` と `month` を使うので十分だと思います。

sjabcdefin replied:
```startDate```からでないと参照できない値ではないため、```month```を使用するように修正いたしました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  if (date.getDay() === 6) {
+    process.stdout.write("\n");
+    numberOfLines++;
+  } else {
+    process.stdout.write(" ");
+  }
```
月末が土曜日でない場合に日付の後ろに余分な空白が出力されます。

sjabcdefin replied:
確認が足りておらず、失礼いたしました。
判定に、```|| date.getDate() === endDate.getDate()```を追加し、月末の場合は改行を出力するように修正いたしました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  }
+}
+
+for (let lines = numberOfLines; lines < 6; lines++) {
```
`lines` だと複数形なので配列が格納されている変数に見えるし、`lines` という名前が何を指しているのか自体も曖昧な気がします。

sjabcdefin replied:
```lines```を```lineBreakCount```に修正いたしました。

sjabcdefin committed:
Update: Modify the argument used when creating the Date object set to the endDate variable.

endDate変数に設定されるDateオブジェクト作成時に使用される引数を修正。

sjabcdefin committed:
Update: Modify the variable name that represents the number of line break.

改行コード数を表す変数名を変更。

sjabcdefin committed:
Update: Modify so that the blank behind the last day of month are not output.

月末が土曜日以外の場合に、日付の後ろの余分な空白が出力されないように修正。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  date <= endDate;
+  date.setDate(date.getDate() + 1)
+) {
+  process.stdout.write(`${String(date.getDate()).padStart(2, " ")}`);
```
文字列中に変数を埋め込んでいるわけではないのにテンプレート文字列が使われています。

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/11687d70ee5c02b9351b53befc44bd6963c77f91

テンプレート文字列を削除し、以下のように修正いたしました。
```process.stdout.write(String(date.getDate()).padStart(2, " "));```

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+console.log("日 月 火 水 木 金 土");
+process.stdout.write("   ".repeat(startDate.getDay()));
+
+let lineBreakCount = 0;
```
この変数の命名についてはコメントしていませんよ。変更してはいけないわけではありませんが、どうして変更することにしたのか教えてほしいです。

sjabcdefin replied:
すみません。
```lineBreakCount```の方が、```numberOfLines```より改行数、改行カウンタを示しているのが直感的に分かりやすいと思い、```lineBreakCount```に修正いたしました。

```numberOfLines```を```lineBreakCount```に修正し、```lines```を削除した経緯は以下となります。

0.  https://github.com/sjabcdefin/js-practices/pull/2#discussion_r1895757985 で指摘いただいたコード
```for(let lines = numberOfLines; lines < 6; lines++) {```

2. 最初、変数```lines```を```LineBreakCount```に修正
```for(let LinBreakCount = numberOfLines; lineBreakCount < 6; lineBreakCount++) {```

3. for文の初期化ブロックでは、```LineBreakCount```に```numberOfLines```を設定する必要はなく、```numberOfLines```をそのまま使用すればよいということに気づき、修正しようと考える。
その中で、```LineBreakCount```と```numberOfLines```どちらを使用するのが適切か考え、```LineBreakCount```の方が、改行数、改行カウンタを示しているのが直感的に分かりやすいと思い、```LineBreakCount```に修正
```for(lineBreakCount; lineBreakCount < 6; lineBreakCount++) {```

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  }
+}
+
+for (lineBreakCount; lineBreakCount < 6; lineBreakCount++) {
```
意味のない初期化処理が書かれています。

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/js-practices/pull/2/commits/88ad0dc8783db016ea6a124fce8ddd0112e86584

for文の初期化ブロックについて、空文に修正いたしました。
```
for (; lineBreakCount < 6; lineBreakCount++) {
  process.stdout.write("\n");
}
```

sjabcdefin committed:
Remove unnecessary template literal in date output

文字列中に変数を埋め込んでいるわけではないため、不要なテンプレート文字列を削除

sjabcdefin committed:
Remove initialization in for-loop

シンプルなコードとするため、意味のない初期化処理を削除
